### PR TITLE
adding documentation for installing on GAE

### LINKED ISSUE #1379: Deployment guide for Google appengine
I would like to deploy athens on Google appengine to handle my private modules. I would prefer a guide for the appengine standard environment. 

**Additional context**
https://twitter.com/arschles/status/1173661298963271680

### PR DESCRIPTION
**Pull request by schafer14**
**What is the problem I am trying to address?**

Documentation/demonstration for installing Athens on GAE

**How is the fix applied?**

Append documentation page to install folder

**Mention the issue number it fixes or add the details of the changes if it doesn't have a specific issue.**

Fixes #1379

### CODE REVIEW

**Comment by schafer14** [comment]
@arschles @tamalsaha 
This is a draft pr for a documentation page on how to install Athens on GAE for an issue @tamalsaha raised. I have a solution that is working, but it involves creating a couple of files (which seems like a lot of overhead compared to other deployment solutions). 

Can any one think of a better solution? It would be optimal to be able to deploy from a single command. 

Thanks.

File: docs/content/install/install-on-gae.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+draft: false
+weight: 4
+---
+
+
+[Google App Engine (GAE)](https://cloud.google.com/appengine/) is a Google service allows applications to be deployed without provisioning the underlying hardware. It is similar to Azure Container Engine which is covered in a [previous section](/install/install-on-aci). This guide will demonstrate how you can get Athens running on GAE.
+
+## Selecting a Storage Provider
+
+Athens currently supports a number of storage drivers. For quick and easy use on GAE, we recommend using the local disk provider. For more permanent use, we recommend using MongoDB or other more persistent infrastructure. For other providers, please see the [storage provider documentation](/configuration/storage).
```
Comment (arschles) [inline docs/content/install/install-on-gae.md]: @schafer14 would it make sense to use GCS for storage on app engine? Not sure if that's quick and easy enough. What do you think?

Comment (tamalsaha) [reply docs/content/install/install-on-gae.md]: It will be very unusual to use Appengine with something other than GCS.

**Comment by tamalsaha** [comment]
Thanks for writing this up. But we ( I assume other people, too) would like a guide that uses Appengine Standard environment. The main benefit of standard environment over the Flexible env (needed for Docker image based deployments) is that standard env provide free daily quota. For lightly used services, this means virtually no bill.

- https://cloud.google.com/appengine/quotas#Instances

To use standard quota, users need a `app.yaml` file and run `gcloud app deploy` which deploys a standard GO net/http server. Here is full working example:
https://github.com/GoogleCloudPlatform/golang-samples/tree/master/appengine/go11x/helloworld

Guide: https://cloud.google.com/appengine/docs/standard/go112/quickstart

Looking at https://github.com/gomods/athens/blob/master/cmd/proxy/main.go , I think standard environment should work with athens. @schafer14 , I wonder if you could give that a try?

**Comment by tamalsaha** [comment]
So, I gave it a shot using standard env. I don't think it will work.
https://github.com/tamalsaha/athens/commit/91a8f15c16e44abe9227bfaa46dbe97a68359a93

Athens seems to shell out to `go` binary. That is not possible with Standard environment. I getting this error:

```
2019-10-02 18:51:38 default[20191002t114523]  2019-10-02 18:51:38.692941 I | Running dev mode with default settings, consult config when you're ready to run in production
2019-10-02 18:51:38 default[20191002t114523]  time="2019-10-02T18:51:38Z" level=info msg="Exporter not specified. Traces won't be exported"
2019-10-02 18:51:38 default[20191002t114523]  2019-10-02 18:51:38.696370 I | error adding proxy routes (exec: "go": executable file not found in $PATH)
```

**Comment by tamalsaha** [comment]
I tried the flexible environment.
https://github.com/tamalsaha/athens/commit/14576f66cf9f41b7b46f53b7e799bf91a57a30eb

It seems to be running: https://goproxy.appscode.dev

File: docs/content/install/install-on-gae.md (hunk 1 above)
Comment (schafer14) [reply docs/content/install/install-on-gae.md]: That does sound good. I'll do that.

**Comment by schafer14** [comment]
@tamalsaha for scale to zero/free tier would Cloud Run be a suitable solution? I completely agree that a freeish way of deployment would be valuable. 

@arschles can I ask a couple of peripherally related questions for my own education: does Azure offer a Knative as a service? Do you know if any one has deployed using Knative/Cloud Run yet?

File: docs/content/install/install-on-gae.md (hunk 1 above)
Comment (schafer14) [reply docs/content/install/install-on-gae.md]: To both config/gae and to gcs.

**Comment by tamalsaha** [comment]
@schafer14 , Cloud Run is the new kid in the block. I see that it is the hotness in Google "serverless" offering. If you are up for it, please add doc for it. I have not had a chance to try it or even look into the docs for it. Does this have a free tier?

**Comment by arschles** [comment]
Sorry for lagging behind on this PR. I'll try to answer everything I missed in here. If I forgot, don't hesitate to ping 😁 

@tamalsaha you're right that standard won't work. We have to shell out to `go` so we don't replicate the VCS (and other) code in the CLI. I'm super duper excited to see that you got the flexible environment working 🎉 

@schafer14 I think cloud run would be great docs also (not instead of) app engine. Those docs would be most welcome! We can add them in this PR, in a follow-up, or of course not at all - it's :100:% your choice

>does Azure offer a Knative as a service?

@schafer14 nope, not at the moment

>Do you know if any one has deployed using Knative/Cloud Run yet?

@schafer14 yup, I know that folks have deployed it on cloud run for internal use. Primarily for serving CI/CD builds

File: docs/content/install/install-on-gae.md (hunk 1 above)
Comment (schafer14) [reply docs/content/install/install-on-gae.md]: @arschles would the scripts directory be a better place for these files then the config file? I don't have a preference, I am just wondering if that is another option.

**Comment by schafer14** [comment]
@tamalsaha yes google cloud run does have a free tier of 50 computer hours per month. I just added documentation for deploy there. 

@arschles @tamalsaha I did make the docs specific for running with the google storage provider. I think you are right that this will most often be the case. 

@arschles for google app engine I am recommending to copy and update a sample configuration. I am not sure that's the best process. What do you think?

File: docs/content/install/install-on-gae.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+There is already a Google Application Engine scaffold set up for you. Copy it into a new file and make changes to the environment variables.
+
+```bash
+cd athens
+cp scripts/gae/app.sample.yaml scripts/gae/app.yaml
+code scripts/gae/app.yaml
+```
+
+Once you have configured the environment variables you can build a GAE service
```
Comment (arschles) [inline docs/content/install/install-on-gae.md]: Would you be ok with making this sentence more specific and saying "... you can deploy Athens as a GAE service." ?

File: docs/content/install/install-on-gae.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+## Selecting a Storage Provider
+
+There is documentaion about how to use environment variables to configure a large number of storage providers; however, for this prarticular example we will use [Google Cloud Storage](https://cloud.google.com/storage/)(GCS) because it fits nicely with Cloud Run.
+
+## Before You Begin
+
+This guide assumes you have completed the following tasks:
+
+- Signed up for Google Cloud
+- Installed the [gcloud](https://cloud.google.com/sdk/install) command line tool
```
Comment (arschles) [inline docs/content/install/install-on-gae.md]: I noticed in the cloud run documents ([below](https://github.com/gomods/athens/pull/1400/files#diff-ec0a7f976eec2a24d58cf6b9b06c5ae6R21)), you say that the user needs to set up a GCS bucket. Do they need to do that here for GAE too?

File: docs/content/install/install-on-google-cloud-run.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+## Setup
+
+First set up configuration environment variables
+
+```bash
+export GOOGLE_CLOUD_PROJECT=your-project
+export ATHENS_REGION=asia-northeast1
+export GCS_BUCKET=your-bucket-name
+```
```
Comment (arschles) [inline docs/content/install/install-on-google-cloud-run.md]: Would you mind making it clear here that the user can change `your-project`, `asia-northeast1`, and `your-bucket-name` to values that make sense for them. What do you think of the below text?

"Change the values of these environment variables to be appropriate for your application. For `GOOGLE_CLOUD_PROJECT`, this needs to be the name of the project that has your cloud run deployment in it. `ATHENS_REGION` should be the region that your cloud run instance will be in (_maybe a link here to the regions list?_), and `GCS_BUCKET` should be the Google Cloud Storage bucket that Athens will store module code and metadata in."

File: docs/content/install/install-on-google-cloud-run.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+You will then need to push a copy of the Athens docker image to your google cloud container registry.
+
+```bash
+docker pull gomods/athens:v0.6.0 
+
+docker tag gomods/athens:v0.6.0 gcr.io/$GOOGLE_CLOUD_PROJECT/athens:v0.6.0
+
+gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/athens:v0.6.0
+```
```
Comment (arschles) [inline docs/content/install/install-on-google-cloud-run.md]: Can you put a `?` character before each command? We do that pattern elsewhere in the documentation to indicate that these are commands that the user should run.

File: docs/content/install/install-on-gae.md (hunk 3 above)
Comment (schafer14) [reply docs/content/install/install-on-gae.md]: Oh gosh, sorry, I had used the same GCS bucket to do both pages and completely forgot to add it here. So glad someone is proofreading!

Comment (schafer14) [reply docs/content/install/install-on-gae.md]: Actually, upon review, it looks like I just straight up forgot to include this yikes.

File: docs/content/install/install-on-google-cloud-run.md (hunk 5 above)
Comment (schafer14) [reply docs/content/install/install-on-google-cloud-run.md]: @arschles I am a little confused about this. Do you mean the line should be? 

```console
? gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/athens:v0.6.0
```

File: docs/content/install/install-on-gae.md (hunk 2 above)
Comment (schafer14) [reply docs/content/install/install-on-gae.md]: Yes much better!

File: docs/content/install/install-on-google-cloud-run.md (hunk 4 above)
Comment (schafer14) [reply docs/content/install/install-on-google-cloud-run.md]: I think that is much clearer. I added the list of regions to this page as a result of the previous comment, so I didn't repeat the link. Is that alright?

Comment (schafer14) [reply docs/content/install/install-on-google-cloud-run.md]: Yes, this is much better.

**Comment by schafer14** [comment]
@arschles I updated everything except the ? at the beginning of commands. Can you clarify that? I'm a little confused.

**Comment by arschles** [comment]
@schafer14 doh, I'm sorry - I typed the wrong character. What I meant to say was, can you put the `$` character at the beginning of each command line. So your example would look like this:

```console
$ gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/athens:v0.6.0
```

**Pull request merged**